kratos/spaces/ublas_space.h

```diff
@@ -710,8 +710,8 @@ class UblasSpace
     {
         const std::size_t system_size = rA.size1();
 
-        const double* Avalues = rA.value_data().begin();
-        const std::size_t* Arow_indices = rA.index1_data().begin();
+        const auto& Avalues = rA.value_data();
```

Comment:
That's a good point, I can imagine writing to the A matrix in a parallel loop, while using the underlying `value_data`, could still lead to issues. I think this needs to be investigated further.

However, in our test cases we saw that the re-allocation consistently caused a segfault/memory acces violation and with this change it consistently passes (since at least the reference is not invalidated like the double* there was before).

Therefore, I'd propose to merge this PR for now, since it fixes our pipelines (which are blocking our development at the moment), and afterwards there can be an investigation into whether we need other changes as well to make this part more secure/threadsafe.

Reply:
Line 737:
```cpp
rA(Index, Index) = scale_factor;
```

I don't see how this would trigger a reallocation. Also, `Index` is unique in each iteration so writing to that component is threadsafe.

I think there might be something fishy going on when iterators are converted to raw pointers.

Reply:
> Line 737:
> 
> ```c++
> rA(Index, Index) = scale_factor;
> ```
> 
> I don't see how this would trigger a reallocation. Also, `Index` is unique in each iteration so writing to that component is threadsafe.
> 
> I think there might be something fishy going on when iterators are converted to raw pointers.

I think if we add more data to the matrix, the 'data_value' array might be updated. Since I don't think it reserves the length of the full matrix, changing the matrix could mean the unbounded_array will be re-allocated.

In any case converting the iterator to a raw pointer is indeed tricky, as it could work depending on the implementation of the begin() iterator, but it could just as well not work. In any case, I think using the references like proposed in this PR is a step in the right direction

Reply:
> `rA(Index, Index) = scale_factor;` with a sparse matrix may allocate if this entry is missing

The diagonal values are guaranteed to exist, check the algorithm of the non zero values.

Reply:
> The diagonal values are guaranteed to exist, check the algorithm of the non zero values.

can you link it?
I cannot see it in the code

Reply:
> adds diagonal terms when the row is empty

thats also how I read it, and in which case the segfault can be explained

Reply:
@RiccardoRossi Just to re-iterate, the current state is blocking the development of the geomechanics team. This PR at least gets rid of the implicit conversion from iterator to pointer and fixes the pipelines at our end. Therefore, can we merge this PR and investigate more in-depth at a later moment by creating a separate issue?

Reply:
well just to say that my proposal would be to use the code

~~~cpp
if(empty)
{
   auto p = rA.find_element(i,i);
   if(p)
       *p = ScalingFactor;
}
~~~

admittedly this is a change of behaviour of the current code, but the current code is wrong...

Reply:
@rfaasse if the insertion in parallel is the true cause of this segfault, then this PR doesn't fix it and your pipeline just dodged the race condition by chance. Merging this will not change that.

If you want to merge this PR as quickly as possible regardless (I'm fine with that), please open that separate issue and link this PR so the problem won't be forgotten.

Reply:
> @RiccardoRossi Just to re-iterate, the current state is blocking the development of the @KratosMultiphysics/geomechanics team. This PR at least gets rid of the implicit conversion from iterator to pointer and fixes the pipelines at our end. Therefore, can we merge this PR and investigate more in-depth at a later moment by creating a separate issue?

hi @rfaasse i approved the PR, nevertheless let me tell you that this is not a fix. 

the problem is that if one of the threads reallocates the matrix (by doing rA(i,i) in a case in which it does not exist) the other threads will have a conflict as the reallocation will happen while they are writing. This is possibly even worst to find than the current behaviour ...

Reply:
sorry to be so verbose, this is just to copy here how the find_element works, because at that point we can also do only what is needed directly in code

~~~cpp
const_pointer find_element (size_type i, size_type j) const {
            size_type element1 (layout_type::index_M (i, j));
            size_type element2 (layout_type::index_m (i, j));
            if (filled1_ <= element1 + 1)
                return 0;
            vector_const_subiterator_type itv (index1_data_.begin () + element1);
            const_subiterator_type it_begin (index2_data_.begin () + zero_based (*itv));
            const_subiterator_type it_end (index2_data_.begin () + zero_based (*(itv + 1)));
            const_subiterator_type it (detail::lower_bound (it_begin, it_end, k_based (element2), std::less<size_type> ()));
            if (it == it_end || *it != k_based (element2))
                return 0;
            return &value_data_ [it - index2_data_.begin ()];
        }
~~~

Reply:
I think there are actually two related problems here. One is that the double* is invalidated upon re-allocation of the underlying vector, regardless of parallel vs serial. By using a reference at least we are still pointing to the right memory for a serial loop when the vector has re-allocated. My guess is that that was the issue for our pipeline, since it was consistently failing before and consistently passing with the fix (but that is just an assumption). 

The second issue is as you all rightly pointed out that this implementation is not thread-safe, for which I'll create an issue and link to this PR and this discussion. Thank you for all the input and will see you in the next one!

Reply:
Interesting stuff

There should be no reallocation any more once its fixed properly. I'd be curious if it would still fail in your pipelines after reverting this PR

Can you please include me in the next PR?